src/rdb/rdb_internal.h

```diff
@@ -91,9 +100,12 @@ struct rdb {
 	uint64_t		d_compact_thres;/* of compactable entries */
 	ABT_cond		d_compact_cv;	/* for triggering base updates */
 	ABT_cond		d_compacted_cv;	/* for d_lc_record.dlr_aggregated updates */
+	ABT_cond                d_checkpoint_cv; /* for triggering checkpointing */
+	ABT_cond                d_commit_cv;     /* for waking active checkpoint */
 	bool			d_stop;		/* for rdb_stop() */
 	ABT_thread		d_timerd;
 	ABT_thread		d_callbackd;
+	ABT_thread              d_checkpointd;
```

Comment:
[Nit] Sorry Jeff, but would you mind putting the new field together, next to each other, after `d_pool`, please? We group fields by their function, not type. Also, the checkpointing is not specifically related to `raft`, but related to pool used by `rdb`.

Accordingly, the new code added to `rdb_raft.c` (including `rdb_checkpointd`) would fit better in `rdb.c`, with prefix `rdb_` instead of `rdb_raft_`. For instance, the checkpoint fields and ULT could be initialized in `rdb_open_internal`. Then, checkpointing will be available even if an RDB is _opened_ (e.g., by Catastrophic Recovery) but not _started_. This would be more logical and enable future CR to write a larger amount of data with checkpointing, if necessary.

May need to add a new mutex together with the new CVs---`d_raft_mutex` and `d_mutex` are both unsuitable for checkpointing purpose, I think. The new checkpoint code should share no fields with existing `rdb` code, except for `d_pool`.

Hope my expression is understandable. :) If not, let's discuss offline.

Reply:
Sure, the PR failed anyway so I can make these changes.